Added more detailed sample of usage

m1l4n54v1c opened the pull request:
Resolves #1

m1l4n54v1c committed:
Added more detailed sample of usage.

Narvalex commented on src/test/java/io/event/thinking/eventstore/sample/SubscriptionCommandHandler.java:
```diff
@@ -0,0 +1,65 @@
+                     // build the command model based on sourced events
+                     .reduce(new Subscription(), this::applyEvent)
+                     // execute the command
+                     .flatMap(Subscription::subscribe)
```
I aso prefer emitting events in the command handler rather than within the 'aggregate,' 😊 but the comment suggests that this method executes the command. However, the subscribe method seems more like a validation call to me.

m1l4n54v1c replied:
You are right! Execution is validation + event publication. Maybe a better name for the method would be `validateSubscription`? I was on the verge of moving the `Subscription` command model to the command handler since `Subscription` feels a bit anemic. However, I wanted to remove any `event store` references from the `Subscription`.

Narvalex replied:
Would returning events to the caller be a possible solution?

m1l4n54v1c replied:
That might be the cleanest solution, after all!

Narvalex commented on src/test/java/io/event/thinking/eventstore/sample/Subscription.java:
```diff
@@ -0,0 +1,61 @@
+
+import reactor.core.publisher.Mono;
+
+class Subscription {
```
This looks like the "aggregate" in the typical event sourced DDD building blocks, but I thought that we will have something like a projected "agregate" on a per command basis, since we are only interested in reading the events neccesary to validate the command and decide something. Wouldnt a better name be somerhing like the command name and a prefix or suffix. Example: SubscribeStudentProjection instead of just "Subscription". Because, how would you name another projection to validate a different command that needs different events to rehydrate and a different set of fields to hold state?

m1l4n54v1c replied:
What's most confusing is the NOUN (forbidden NOUN :)). However, `Subscription` models a process rather than a state. 
One thing we must be careful to avoid doing is to have a command model per command. While in some cases it is a better approach, in others, reusability of the command model and grouping command handlers together would be more sane to do. To give you an example, the `UnsubscribeStudent` command would naturally land in the `Subscription` command model since it requires the same events; the model is built in the same way - we'd like to check whether student and course exist, whether number of courses per student is > 0, number of students per course is > 0, and we don't want to unsubscribe the student from the course one hasn't subscribed to.

Anyway, I'm open to suggestions for better naming :)

Narvalex replied:
Here's the thing: To unsubscribe, you only need to check if the student and the course exist, and if the student is currently enrolled in that course. You don’t need to worry about course capacity when unsubscribing—that only matters during the subscription process. What I find beautiful about this approach is the flexibility to create as many models as necessary to validate commands, even having one per command if needed, instead of dealing with those large, complex aggregates that I’m more accustomed to.

m1l4n54v1c replied:
Actually, it's even simpler. We could do this with the criterion containing both tags, for course, and student. I'll build this.

Narvalex reviewed (commented):
Thanks a lot @m1l4n54v1c for your time

m1l4n54v1c committed:
Added more logic to the faculty sample.
Introduced microdcb framework to makes modeling easier.

m1l4n54v1c committed:
Re-organized modules.

m1l4n54v1c committed:
Prevent instantiation of constants.

m1l4n54v1c committed:
Added support for a transaction of events to the event store.
Changed the return type of handling the command in command model since we need a finite list of events to form a transaction.

m1l4n54v1c committed:
Split store module to event-store-api and inmemory implementation.

m1l4n54v1c committed:
Renamed buildCriteria in command model.

m1l4n54v1c committed:
Sourcing the events is not eager anymore.

m1l4n54v1c committed:
Added possibility to append a transaction of events.
javadoc.

m1l4n54v1c committed:
Added tests for Unsubscription use-case.
